### ISSUE TITLE
Configure Dex with CRDs

### ISSUE THREAD

**Issue by logicfox** [dexidp/dex#1381]
I haven't found any examples of configuring Connectors and OAuth2Clients via CRDs. It would be useful to know how to do this, or add missing support, as this would enable users of an RBAC enabled cluster to configure Dex simply with CRDs.

For instance, I tried creating an OIDC connector based on [this struct](https://github.com/dexidp/dex/blob/master/storage/kubernetes/types.go#L660-L672)
```yaml
apiVersion: dex.coreos.com/v1
kind: Connector
metadata:
  namespace: dex
  name: testconnector
id: atest
type: oidc
name: atest
config:
  issuer: "https://issuer.url"
  clientID: "the-client-id"
  clientSecret: "oh-so-secret"
  redirectURI: "https://dex.application.net/callback/"
```
but Dex fails while starting up with the error message from [this line](https://github.com/dexidp/dex/blob/v2.13.x/storage/kubernetes/storage.go#L395)

```
failed to initialize server: server: failed to list connector objects from storage: failed to list connectors: json: cannot unmarshal object into Go struct field Connector.config of type []uint8
```

**Comment by srenatus** [comment]
Yup, looks like something where some more docs would be helpful, indeed. There's [docs](https://github.com/dexidp/dex/blob/master/Documentation/storage.md#kubernetes-custom-resource-definitions-crds), but they could surely be expanded with some config examples.

Does looking at [the helm chart](https://github.com/helm/charts/tree/master/stable/dex) help in any way? I'm sorry I've never messed with CRDs much...

**Comment by mrwulf** [comment]
@logicfox, I agree that it would be really convenient to be able to define clients with CRDs, but this doesn't currently work -- as the [documentation](https://github.com/dexidp/dex/blob/master/Documentation/storage.md#kubernetes-custom-resource-definitions-crds) states, "Admins should not interact with these resources directly, except while debugging. These resources are only designed to store state and aren't meant to be consumed by end users. For modifying dex's state dynamically see the API documentation."

**Comment by ericchiang** [comment]
> @logicfox, I agree that it would be really convenient to be able to define clients with CRDs, but this doesn't currently work -- as the documentation states, "Admins should not interact with these resources directly, except while debugging. These resources are only designed to store state and aren't meant to be consumed by end users. For modifying dex's state dynamically see the API documentation."

Yes, this is unsupported.

**Comment by MatteoJoliveau** [comment]
Hi! Sorry to bump an old issue, I just wanted to add a use-case to this topic: GitOps

When following GitOps practices the state of a system is defined in terms of files in Git instead of being imperatively driven via API requests. Having supported access to Dex's CRDs would allow us to define clients in a way that is decoupled from the Dex deployment files (which may live in a separate repository).

Example:

I have an `infrastructure` repository which holds manifests for Dex and a bunch of other infra-related services (load balancers, Grafana etc)

One developer wants to deploy an internal custom webapp which should authenticate users with the same user pool as the rest of the infra services. Instead of having to edit Dex's configuration file in the `infrastructure` repository, they could simply add a manifest for a `kind: OAuth2Client` alongside the rest of the webapp deployment to have it automatically created by Dex.

One example of Kubernetes-enabled system that works this way is [ArgoCD](https://argoproj.github.io/argo-cd/operator-manual/declarative-setup/) which supports creating apps and projects via either a web UI, a gRPC API or by applying custom resources.

Given that Dex already has its own CRDs, I guess supporting this use-case could simply mean adding a validation schema to the CRDs and documenting how to create them via YAML files.

**Comment by stefansedich** [comment]
> Hi! Sorry to bump an old issue, I just wanted to add a use-case to this topic: GitOps
> 
> When following GitOps practices the state of a system is defined in terms of files in Git instead of being imperatively driven via API requests. Having supported access to Dex's CRDs would allow us to define clients in a way that is decoupled from the Dex deployment files (which may live in a separate repository).
> 
> Example:
> 
> I have an `infrastructure` repository which holds manifests for Dex and a bunch of other infra-related services (load balancers, Grafana etc)
> 
> One developer wants to deploy an internal custom webapp which should authenticate users with the same user pool as the rest of the infra services. Instead of having to edit Dex's configuration file in the `infrastructure` repository, they could simply add a manifest for a `kind: OAuth2Client` alongside the rest of the webapp deployment to have it automatically created by Dex.
> 
> One example of Kubernetes-enabled system that works this way is [ArgoCD](https://argoproj.github.io/argo-cd/operator-manual/declarative-setup/) which supports creating apps and projects via either a web UI, a gRPC API or by applying custom resources.
> 
> Given that Dex already has its own CRDs, I guess supporting this use-case could simply mean adding a validation schema to the CRDs and documenting how to create them via YAML files.

We would love this too, our current story for provisioning new apps onto Dex is a bit painful.

**Comment by MatteoJoliveau** [comment]
Sorry to nudge this issue again, but declarative management of clients and providers is something really we really need in order to integrate Dex in our environment. I would like to ask the maintainers if they have plans for this, or if they want to delegate this to an third party "Dex Operator" project.

IMHO, since Dex already supports CRDs for configuration it would be only natural to use them and promote them to the status of "public API", compared to the current "internal implementation detail".

**Comment by sagikazarmark** [comment]
We'll add this to the roadmap at some point, but can't promise anything or give an ETA at this point. We are still figuring things out around the CNCF sandbox status, trying to clean up our processes, etc. But I think this is definitely the way Dex needs to head, or at least we need something similar to Prometheus for configuration.

**Comment by bonifaido** [comment]
> IMHO, since Dex already supports CRDs for configuration it would be only natural to use them and promote them to the status of "public API", compared to the current "internal implementation detail".

Dex CRDs are used for **storage** and not for **configuration** that is why they are an internal implementation detail.

**Comment by MatteoJoliveau** [comment]
@bonifaido I know, that's what I'm saying. The fact that Dex already has CRDs means it should not be too complicated to promote them from internal storage implementation to an officially supported way of configuration, aligning Dex's user experience to other projects in the kube world (namely the Argo projects, the operator pattern, etc) and enabling fully declarative Dex deployments

**Comment by sagikazarmark** [comment]
The fact that it uses CRDs for *something* (storage in this case) doesn't mean that configuring Dex via CRDs is two clicks away. 🤣

**Comment by MatteoJoliveau** [comment]
I didn't mean to trivialize the task. I just meant to point out that it's not like Dex has to be integrated from scratch, so I wanted to encourage maintainers to consider this feature.
I'm really happy it's on the roadmap, we'll implement some temporary wrapper while this is getting built to achieve our goals.

**Comment by harmjanblok** [comment]
Based on the example of @logicfox and reverse engineering of some code, I was able to define a connector with following CRD:

```yaml
apiVersion: dex.coreos.com/v1
kind: Connector
metadata:
  namespace: dex
  name: atest # id of your connector
id: atest # should match metadata.name
type: oidc
name: 'A test'
# config, as described on https://dexidp.io/docs/connectors/oidc/, converted to json, base64 encoded
config: eyJpc3N1ZXIiOiJodHRwczovL2lzc3Vlci51cmwiLCJjbGllbnRJRCI6InRoZS1jbGllbnQtaWQiLCJjbGllbnRTZWNyZXQiOiJvaC1zby1zZWNyZXQiLCJyZWRpcmVjdFVSSSI6Imh0dHBzOi8vZGV4LmFwcGxpY2F0aW9uLm5ldC9jYWxsYmFjay8ifQo=
```

`metadata.name` needs to be encoded for Passwords and OAuth2Clients, see comment on https://github.com/dexidp/dex/issues/1606#issuecomment-569130595.

**Comment by mwmahlberg** [comment]
> config, as described on https://dexidp.io/docs/connectors/oidc/, converted to json, base64 encoded

Apprently, it does not have to be converted to json. Works fine with me as base64 encoded yaml.

I _think_ "all" that would have to be done is to replace

https://github.com/dexidp/dex/blob/205a7a43a551fbae5f48980eddd1b5be1dd53784/storage/kubernetes/types.go#L711

with the actual struct as in

https://github.com/dexidp/dex/blob/205a7a43a551fbae5f48980eddd1b5be1dd53784/connector/github/github.go#L41

For each connector type - so there would have to be a `dex.coreos.com/v1/GitHubConnector` etc instead of a `dex.coreos.com/v1/Connector`.

This comes with some ramifications, however: Should the actual Version be incremented? If not, how is compatibility maintained? I could think of taking the `dex.coreos.com/v1/Connector`, decoding the config field, identify the connector and then create a `dex.coreos.com/v1/<XY>Connector` resource from it. However, that again comes with ramifications. I am not sure I'd want the actual IDP to be able to create such things as completely new connectors - actually I'd want it to be the last service to be able to do so: Any hijacker can wreak havoc if able to basically add users and Auth Methods.

My suggestion would be to simply add the `<XY>Connector` CRDs. I will dig into that.

**Comment by petr0ff** [comment]
> Based on the example of [@logicfox](https://github.com/logicfox) and reverse engineering of some code, I was able to define a connector with following CRD:
> 
> apiVersion: dex.coreos.com/v1
> kind: Connector
> metadata:
>   namespace: dex
>   name: atest # id of your connector
> id: atest # should match metadata.name
> type: oidc
> name: 'A test'
> # config, as described on https://dexidp.io/docs/connectors/oidc/, converted to json, base64 encoded
> config: eyJpc3N1ZXIiOiJodHRwczovL2lzc3Vlci51cmwiLCJjbGllbnRJRCI6InRoZS1jbGllbnQtaWQiLCJjbGllbnRTZWNyZXQiOiJvaC1zby1zZWNyZXQiLCJyZWRpcmVjdFVSSSI6Imh0dHBzOi8vZGV4LmFwcGxpY2F0aW9uLm5ldC9jYWxsYmFjay8ifQo=
> `metadata.name` needs to be encoded for Passwords and OAuth2Clients, see comment on [#1606 (comment)](https://github.com/dexidp/dex/issues/1606#issuecomment-569130595).

yep, thanks, that works. But looks like it cannot work with vault + external secrets operator + secretsFromEnv setup, like in Dex's secret/configMap I have say bindPW: $LDAP_PASSWORD (this env lives in dex-secrets which ESO pulls from vault) and Dex deployment has this env "mounted" from dex-secrets.

**Comment by R-Studio** [comment]
Any news?

**Comment by ghf20** [comment]
I have a controller and CR’s that we’ve been using in production for the last 8 months. It allows us to use secrets from ESO for our static clients and connectors in the CM. Happy to contribute it upstream. The main use case was to provision SSO apps with a helm chart without ever needing to manually change the CM

**Comment by mikejoh** [comment]
> I have a controller and CR’s that we’ve been using in production for the last 8 months. It allows us to use secrets from ESO for our static clients and connectors in the CM. Happy to contribute it upstream. The main use case was to provision SSO apps with a helm chart without ever needing to manually change the CM

@ghf20 Sounds fantastic! Is that hosted on GitHub or somewhere public by any chance? I’d be very interested to see how you handled the implementation!
